### DISCUSSION TITLE
Hmac header validation

### DISCUSSION THREAD

**Discussion by forrestab** [dotnet/yarp#2129] [Q&A] [answered]
I am considering a single gateway that will support both api and webhook routes. For the webhook routes I would like to validate the signature header using hmac before proxying to a service ([example](https://confluence.atlassian.com/bitbucketserver/manage-webhooks-938025878.html#Managewebhooks-webhooksecretsSecuringyourwebhook)).

Is a middleware the best place to put this validation? If so, is there a way I can configure the middleware for specific routes or do I need to add some conditions for those specific routes (i.e. the webhook routes)?

**Comment by MihaZupan** [comment, marked as answer]
If you already have a YARP route specific to webhooks, you can add a transform to just that route
```c#
.AddTransforms(builder =>
{
    if (builder.Route.RouteId == "myWebhookRoute")
    {
        builder.AddRequestTransform(async context =>
        {
            context.HttpContext.Request.EnableBuffering();

            if (false) // your validation
            {
                // As you're writing to the response, the proxy step will be skipped.
                context.HttpContext.Response.StatusCode = StatusCodes.Status400BadRequest;
                await context.HttpContext.Response.WriteAsync("Nope");
            }
        });
    }
});
```

Alternatively in a middleware
```c#
app.MapReverseProxy(app =>
{
    app.Use(async (context, next) =>
    {
        if (context.GetReverseProxyFeature().Route.Config.RouteId == "myWebhookRoute")
        {
            // ...
        }

        await next();
    });

    app.UseSessionAffinity();
    app.UseLoadBalancing();
    app.UsePassiveHealthChecks();
});
```
